### ISSUE TITLE
Problems with Angular CLI 6

### ISSUE THREAD

**Issue by binarious** [pnpm/pnpm#1260]
### Code to reproduce the issue:
```
npm i -g @angular/cli
ng new test-pnpm --skip-install
cd test-pnpm
pnpm i
ng build --prod
```

### Expected behavior:
Working build.

### Actual behavior:
```
Versions of @angular/compiler-cli and typescript could not be determined.
The most common reason for this is a broken npm install.

Please make sure your package.json contains both @angular/compiler-cli and typescript in
devDependencies, then delete node_modules and package-lock.json (if you have one) and
run npm install again.
```

### Additional information:

 - `node -v` prints: v10.6.0
 - Windows, OS X, or Linux?: macOS 10.13.5

`@angular/compiler-cli` is defined in the package.json.

**Comment by IpiVasquez** [comment]
I'd like to add: same behavior with Linux (node v10.6.0). We'll be waiting for a fix. In the mix while I have to stick with npm.

**Comment by PaulAGH** [comment]
Same behavior on Windows 7 (node v8.11.3).

**Comment by ThomasSchoenbeck** [comment]
i am running into a similar problem with ng serve.

tried installing typescript globally as some older posts are suggesting. also tried --shamefully-flatten. both did not work.

only solution for me until now is going back to npm install :(.

**Comment by binarious** [comment]
They're defined as peerDependencies in their package.json.

Am 14. Oktober 2018 02:22:45 MESZ schrieb Jason <notifications@github.com>:

**Comment by seangwright** [comment]
@JasonGoemaat 
> Shouldn't the `package.json` for that package have a dependency for `@angular/compiler-cli`? Maybe some of the tools don't use dependencies assuming they can find the packages because they are installed elsewhere normally?

It's listed as a `devDependency` and a `peerDependency` since the Angular v7 betas, but before that it wasn't listed at all.

- Included in 7.0.1
https://github.com/angular/angular-cli/blob/v7.0.1/packages/ngtools/webpack/package.json

- Missing in 6.2.7
https://github.com/angular/angular-cli/blob/v6.2.7/packages/ngtools/webpack/package.json

**Comment by marcelcremer** [comment]
I just ran into the same issue with Angular 6. From what I understand in this thread, Angular doesn't have correct dependencies set in their packages?

How comes that NPM resolves those deps, and pnpm doesn't? If the package.json is wrong, is pnpm aiming to behave the same as npm does?

I'd love to provide more infos on this if needed, or try to help otherwise if I can :)

**Comment by zkochan** [comment]
Sorry that I am not digging into it but after maintaining pnpm for more than 2 years I am sure pnpm works correctly in all cases. Some version of Angular worked with pnpm but then, unfortunately, things went south.

These are the issues that may happen:

1. packages require other packages that are not declared in their `package.json`. This works with flat node_modules but does not with pnpm's non-flat node_modules (https://medium.com/pnpm/pnpms-strictness-helps-to-avoid-silly-bugs-9a15fb306308)
1. toolings try to reimplement Node's module resolution algorithm but do it incorrectly by preserving symlinks (Node does not preserve symlinks, that is why it works with pnpm's [symlinked node_modules](https://medium.com/pnpm/flat-node-modules-is-not-the-only-way-d2e40f7296a3))

In most cases these issues are solved when installing a flat node_modules. You can do that with pnpm by setting `shamefully-flatten=true` in an `.npmrc` file in the root of your project.

if Angular still doesn't work with `shamefully-flatten=true`, let me know.

**Comment by marcelcremer** [comment]
Thank you for your answer. I've checked my repository, and shamefully-flatten makes it better. However, with my cli version, #767 hits me, which is caused by an old dependency from the angular cli. 

It looks like there are too many parties involved, that would need to fix things:
- angular would need to fix their dependencies (shamefully-flatten workaround available)
- angular-cli needs to upgrade the webpack-license-plugin (resolving strategy was fixed with https://github.com/xz64/license-webpack-plugin/issues/42), which could take a while
- webpack needs to fix their resolval strategy (https://github.com/webpack/webpack/issues/5087 and #801)
- webpack loader problems with symlinks (first post #1262)
- maybe others...

Now that I think I understand, that most problems come from the outside, I sadly switched back to npm for angular :-/

P.S.: Nevertheless, pnpm saves me about 50% time for every build I make for our nodejs-apps with Gitlab-CI. Just wanted to also give some props and <3 to @zkochan and the pnpm contributors. Thank you guys :-)
